python/graphstorm/dataloading/sampler.py

```diff
@@ -210,3 +210,47 @@ def _gen_neg_pair(pos_pair, utype, vtype):
             pos_neg_tuple = _gen_neg_pair(pos_pairs,
                 g.canonical_etypes[0][0], g.canonical_etypes[0][2])
         return pos_neg_tuple
+
+class JointLocalUniform(JointUniform):
+    '''Jointly corrupt a group of edges.
+
+    The main idea is to sample a set of nodes and use them to corrupt all edges in a mini-batch.
+    This algorithm won't change the sampling probability for each individual edge, but can
+    significantly reduce the number of nodes in a mini-batch.
+    The difference between JointUniform and JointLocalUniform is that JointUniform sample
+    negative nodes from the entire graph, but JointLocalUniform only sample negative nodes
+    from the local partition.
+
+    Parameters
+    ----------
+    k : int
+        The number of negative examples per edge.
+    '''
+    def __init__(self, k, per_trainer=False):
+        self._local_neg_nids = {}
+        self._per_trainer = per_trainer
+        super(JointLocalUniform, self).__init__(k)
+
+    def _generate(self, g, eids, canonical_etype):
+        _, _, vtype = canonical_etype
+        shape = eids.shape
+        dtype = eids.dtype
+        ctx = eids.device
+        src, _ = g.find_edges(eids, etype=canonical_etype)
+        src = np.repeat(src.numpy(), self.k)
+
+        if vtype not in self._local_neg_nids:
+            pb = g.get_partition_book()
+            if self._per_trainer:
+                neg_idx = node_split(th.full((g.num_nodes(vtype),), True, dtype=th.bool),
+                                     pb, ntype=vtype, force_even=False,
+                                     node_trainer_ids=g.nodes[vtype].data['trainer_id'])
+            else:
+                neg_idx = node_split(th.full((g.num_nodes(vtype),), True, dtype=th.bool),
```

Comment:
do we need to store them?

Reply:
Yes. Otherwise we need to compute it every time we do negative sampling.

Reply:
oh, both branches prepare negative sampling for the local partition. one is to sample from the local machine, the other is to sample from the local trainer. can we simplify this? do we really need to sample from the local trainer? sampling from the local machine should be enough.